msal/src/main/java/com/microsoft/identity/client/controllers/LocalMSALController.java

```diff
@@ -2,63 +2,127 @@
-        authRequest.setScope(StringUtil.join(' ', request.getScopes()));
+        //2) Gather authorization interactively
+        AuthorizationResult result = performAuthorizationRequest(oAuth2Strategy, parameters);
 
+        if(result.getAuthorizationStatus().equals(AuthorizationStatus.SUCCESS)){
+            //3) Exchange authorization code for token
+            TokenResult tokenResult = performTokenRequest(oAuth2Strategy, result.getAuthorizationResponse(), parameters);
+            if(tokenResult != null && tokenResult.getSuccess()){
+                //4) Save tokens in token cache
+                //saveTokens(oAuth2Strategy, getAuthorizationRequest(oAuth2Strategy, parameters), tokenResult.getTokenResponse(), parameters.getTokenCache());
+            }
+        }
 
-        //TODO: Replace with factory to create the correct Authorization Strategy based on device capabilities and configuration
-        mAuthorizationStrategy = new DeviceBrowserAuthorizationStrategy();
+        throw new UnsupportedOperationException();
+    }
+
+    private AuthorizationResult performAuthorizationRequest(OAuth2Strategy strategy, MSALAcquireTokenOperationParameters parameters) throws ExecutionException, InterruptedException {
 
-        Future<AuthorizationResult> future = mOAuthStrategy.requestAuthorization(authRequest, mAuthorizationStrategy);
+        //TODO: Replace with factory to create the correct Authorization Strategy based on device capabilities and configuration
+        mAuthorizationStrategy = new DeviceBrowserAuthorizationStrategy(strategy, parameters.getActivity());
 
-        future.get();
+        Future<AuthorizationResult> future = strategy.requestAuthorization(getAuthorizationRequest(parameters), mAuthorizationStrategy);
 
-        //We could implement Timeout Here if we wish instead of looping forever
+        //We could implement Timeout Here if we wish instead of blocking indefinitely
         //future.get(10, TimeUnit.MINUTES);  // Need to handle timeout exception in the scenario it doesn't return within a reasonable amount of time
-        //AuthorizationResult authorizationResult = future.get();
+        AuthorizationResult result = future.get();
+
+        return result;
+
+    }
+
+    private AuthorizationRequest getAuthorizationRequest(MSALAcquireTokenOperationParameters parameters){
+        AuthorizationRequest authRequest = new MicrosoftStsAuthorizationRequest();
+
+        String scopes = StringUtil.join(' ', parameters.getScopes());
 
+        authRequest.setClientId(parameters.getClientId());
+        authRequest.setRedirectUri(parameters.getRedirectUri());
+        authRequest.setScope(scopes);
+        authRequest.setResponseType(AuthorizationRequest.ResponseTypes.CODE);
 
+        return authRequest;
     }
 
+    private TokenResult performTokenRequest(OAuth2Strategy strategy, AuthorizationResponse response, MSALAcquireTokenOperationParameters parameters) {
+
+        TokenRequest tr = new TokenRequest();
```

Comment:
nit: tokenRequest might be a good name for consistency

Reply:
not sure I follow... a TokenRequest is a noun.

Reply:
fixed.  Sorry for the misread of your comment.  Eyes apparently playing tricks on me.